source/shadow-tasks/shadow_device_task.c

```diff
@@ -0,0 +1,815 @@
+                        SHADOW_TOPIC_STRING_UPDATE_DELTA( democonfigCLIENT_IDENTIFIER ) ) );
+        }
+    }
+
+    if( xReturnStatus == true )
+    {
+        xReturnStatus = addSubscription( ( SubscriptionElement_t * ) xGlobalMqttAgentContext.pIncomingCallbackContext,
+                                         SHADOW_TOPIC_STRING_UPDATE_ACCEPTED( democonfigCLIENT_IDENTIFIER ),
+                                         SHADOW_TOPIC_LENGTH_UPDATE_ACCEPTED( democonfigCLIENT_IDENTIFIER_LENGTH ),
+                                         prvIncomingPublishUpdateAcceptedCallback,
+                                         NULL );
+
+        if( xReturnStatus == false )
+        {
+            LogError( ( "Failed to register an incoming publish callback for topic %.*s.",
+                        SHADOW_TOPIC_LENGTH_UPDATE_ACCEPTED( democonfigCLIENT_IDENTIFIER_LENGTH ),
+                        SHADOW_TOPIC_STRING_UPDATE_ACCEPTED( democonfigCLIENT_IDENTIFIER ) ) );
+        }
+    }
+
+    if( xReturnStatus == true )
+    {
+        xReturnStatus = addSubscription( ( SubscriptionElement_t * ) xGlobalMqttAgentContext.pIncomingCallbackContext,
+                                         SHADOW_TOPIC_STRING_UPDATE_REJECTED( democonfigCLIENT_IDENTIFIER ),
+                                         SHADOW_TOPIC_LENGTH_UPDATE_REJECTED( democonfigCLIENT_IDENTIFIER_LENGTH ),
+                                         prvIncomingPublishUpdateRejectedCallback,
+                                         NULL );
+
+        if( xReturnStatus == false )
+        {
+            LogError( ( "Failed to register an incoming publish callback for topic %.*s.",
+                        SHADOW_TOPIC_LENGTH_UPDATE_REJECTED( democonfigCLIENT_IDENTIFIER_LENGTH ),
+                        SHADOW_TOPIC_STRING_UPDATE_REJECTED( democonfigCLIENT_IDENTIFIER ) ) );
+        }
+    }
+
+    /* Store the result in the application defined context so the calling task
+     * can check it. */
+    pxApplicationDefinedContext->xReturnStatus = xReturnStatus;
+
+    xTaskNotifyGive( xShadowDeviceTaskHandle );
+}
+
+/*-----------------------------------------------------------*/
+
+static void prvIncomingPublishUpdateDeltaCallback( void * pxSubscriptionContext,
+                                                   MQTTPublishInfo_t * pxPublishInfo )
+{
+    static uint32_t ulCurrentVersion = 0; /* Remember the latest version number we've received */
+    uint32_t ulVersion = 0UL;
+    uint32_t ulNewState = 0UL;
+    char * pcOutValue = NULL;
+    uint32_t ulOutValueLength = 0UL;
+    JSONStatus_t result = JSONSuccess;
+
+    /* Remove compiler warnings about unused parameters. */
+    ( void ) pxSubscriptionContext;
+
+    configASSERT( pxPublishInfo != NULL );
+    configASSERT( pxPublishInfo->pPayload != NULL );
+
+    LogDebug( ( "/update/delta json payload:%s.", ( const char * ) pxPublishInfo->pPayload ) );
```

Comment:
I don't think MQTT guarantees that the payload is always NULL terminated. It's from a buffer that may have been written in previously